source/PlayerInfo.cpp

```diff
@@ -1470,6 +1517,192 @@ bool PlayerInfo::TakeOff(UI *ui)
+// or bribeable. This will check available cargo, and assumes you will
+// sell other commodities to make room for the best one.
+// 0: No, or no space, or already full. 1: Space in fleet. 2: Space in Flagship only
+int PlayerInfo::HasBestTrade(const System *destination)
+{
+	if(!destination
+		|| destination == system
+		|| !HasVisited(*destination)
+		|| !destination->IsInhabited(Flagship())
+		|| !destination->HasTrade())
+		return 0;
+
+	// See if we can potentially land in the system
+	bool landable = false;
+	for(const StellarObject &object : destination->Objects())
+	{
+		if(object.HasSprite() && object.HasValidPlanet())
+		{
+			const Planet *planet = object.GetPlanet();
+			if(planet->CanLand(*Flagship()))
+			{
+				landable = true;
+				break;
+			}
+			// TODO: optionally skip bribeable planets for auto-buy
+			const string& lang = planet->GetGovernment()->Language();
+			if(planet->GetBribeFraction() &&
+				(lang.empty() || GetCondition("language: " + lang)))
+			{
+				landable = true;
+				break;
+			}
+		}
+	}
+	if(!landable)
+		return 0;
+
+	string type = BestTradeType(*destination);
+	if(type.empty())
+		return 0;
+
+	int sizeForCommodities = cargo.Size()
+		- cargo.OutfitsSize() - cargo. MissionCargoSize() - cargo.Get(type);
+
+	if(sizeForCommodities <= 0)
+		return 0;
+
+	if(sizeForCommodities <= Flagship()->Cargo().Size())
+		return 2;
+
+	return 1;
+}
+
+
+
+// What is the best commodity to trade from the current system to the destination
+string PlayerInfo::BestTradeType(const System &destination)
+{
+	int64_t bestProfit = 0;
+	string type;
+
+	for(const Trade::Commodity &commodity : GameData::Commodities())
+	{
+		int64_t sellPrice = destination.Trade(commodity.name);
+		int64_t purcPrice = system->Trade(commodity.name);
+		int64_t profit = sellPrice - purcPrice;
+		if(profit > bestProfit)
+		{
+			bestProfit = profit;
+			type = commodity.name;
+		}
+	}
+
+	return type;
+}
+
+
+
+// Purchase commodities that make the most profit when sold at destination
+void PlayerInfo::BuyBestTrade(const System &destination, bool includeFlagship, bool sellFirst)
+{
+	autoBoughtType = BestTradeType(destination);
+	if(autoBoughtType.empty())
+		return;
+
+	if(sellFirst)
+	{
+		int64_t profitSave = profit;
+		int tonsSoldSave = tonsSold;
+		SellCommodities(autoBoughtType);
+		int64_t profitAuto = profit - profitSave;
+		int tonsSoldAuto = tonsSold - tonsSoldSave;
+		profit = profitSave;
+		tonsSold = tonsSoldSave;
+
+		if(profitAuto)
+		{
+			// Report if robo-merchant sold things to make room
+			string message = "You sold " + to_string(tonsSoldAuto)
+				+ (tonsSoldAuto == 1 ? " ton" : " tons")
+				+ " of cargo to make room for " + autoBoughtType + " ";
+
+			if(profitAuto < 0)
+				message += "at a loss of " + Format::Credits(-profitAuto) + " credits.";
+			else
+				message += "for a total profit of " + Format::Credits(profitAuto) + " credits.";
+
+			Messages::Add(message, Messages::Importance::High);
+		}
```

Comment:
this is the same logic used at the lines after 1479
I would suggest trying to put it into a method since both are in the same PlayerInfo class.

Reply:
They are slightly different because this one is an automated sell, before launching from the planet, and has the extra clause `to make room for " + autoBoughtType + "`. Also similar to the message for `tons of excess cargo`.

Reply:
That seems like a bool to me (I mean just give a bool to the function; in one case it does the first thing, in the other the second)